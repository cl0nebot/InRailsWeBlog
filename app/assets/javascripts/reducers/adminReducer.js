'use strict';

import {
    Record
} from 'immutable';

import * as ActionTypes from '../constants/actionTypes';

import * as Records from '../constants/records';

const initState = new Record({
    currentId: window.currentAdminId ? parseInt(window.currentAdminId, 10) : undefined,
    isConnected: !!window.currentAdminId,

    admin: undefined,

    errors: undefined
});

export default function adminReducer(state = new initState(), action) {
    switch (action.type) {
        case ActionTypes.ADMIN_FETCH_SUCCESS:
            return state.merge({
                admin: new Records.UserRecord(action.user)
            });

        case ActionTypes.ADMIN_SETTINGS_CHANGE_ERROR:
            return state.merge({
                errors: action.errors
            });

        default:
            return state;
    }
};
