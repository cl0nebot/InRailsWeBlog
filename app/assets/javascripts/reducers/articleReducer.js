'use strict';

import {
    Record,
    Map,
    List
} from 'immutable';

import * as ActionTypes from '../constants/actionTypes';

import * as Records from '../constants/records';

import {
    toList,
    fetchReducer,
    mutationReducer,
    mutateArray
} from './mutators';

const initState = new Record({
    isFetching: false,
    isProcessing: false,
    errors: new Map(),

    articles: new List(),
    pagination: new Map(),

    article: undefined,
    articleEditionId: undefined,
    articleVersions: undefined
});

export default function articleReducer(state = new initState(), action) {
    switch (action.type) {
        case ActionTypes.ARTICLE_FETCH_INIT:
        case ActionTypes.ARTICLE_FETCH_SUCCESS:
        case ActionTypes.ARTICLE_FETCH_ERROR:
            return fetchReducer(state, action, (payload) => {
                if (payload.article) {
                    return {
                        article: new Records.ArticleRecord(payload.article)
                    };
                } else {
                    if (payload.infinite) {
                        return {
                            articles: state.articles.concat(toList(payload.articles, Records.ArticleRecord))
                        };
                    } else {
                        return {
                            articles: toList(payload.articles, Records.ArticleRecord)
                        };
                    }
                }
            }, ['infinite']);

        case ActionTypes.ARTICLE_EDITION:
            return state.merge({
                articleEditionId: action.articleId
            });

        case ActionTypes.ARTICLE_CHANGE_INIT:
        case ActionTypes.ARTICLE_CHANGE_SUCCESS:
        case ActionTypes.ARTICLE_CHANGE_ERROR:
            return mutationReducer(state, action, (payload) =>
                payload.articles ? ({
                    articles: toList(payload.articles, Records.ArticleRecord)
                }) : ({
                    article: payload.article && (new Records.ArticleRecord(payload.article)),
                    articles: mutateArray(state.articles, payload.article && (new Records.ArticleRecord(payload.article)), action.removedId)
                }) /*, ['article']*/);

        // History and restoration
        case ActionTypes.ARTICLE_HISTORY:
            return state.merge({
                articleVersions: action.versions
            });

        case ActionTypes.ARTICLE_RESTORE:
            return state.merge({
                article: new Records.ArticleRecord(action.article),
                articleVersions: undefined
            });

        default:
            return state;
    }
};
