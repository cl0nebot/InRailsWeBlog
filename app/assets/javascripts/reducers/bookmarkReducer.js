'use strict';

import {
    Record,
    List
} from 'immutable';

import * as ActionTypes from '../constants/actionTypes';

import {
    mutateArray
} from './mutators';

const initState = new Record({
    bookmarks: new List(),
    unbookmarks: new List()
});

export default function bookmarkReducer(state = new initState(), action) {
    switch (action.type) {
        case ActionTypes.BOOKMARK_ADD:
            return state.merge({
                bookmarks: mutateArray(state.bookmarks, action.bookmark, null, 'bookmarkedId')
            });
        case ActionTypes.BOOKMARK_DELETE:
            return state.merge({
                bookmarks: mutateArray(state.bookmarks, action.bookmark, action.removedBookmarkId, 'bookmarkedId'),
                unbookmarks: state.bookmarks.concat([action.bookmark])
            });

        default:
            return state;
    }
};
