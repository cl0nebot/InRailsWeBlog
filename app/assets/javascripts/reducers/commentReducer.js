'use strict';

import {
    Record,
    Map,
    List
} from 'immutable';

import * as ActionTypes from '../constants/actionTypes';

import * as Records from '../constants/records';

import {
    toList,
    fetchReducer,
    mutationReducer,
    findItemIndex
} from './mutators';

const initState = new Record({
    // Required for fetchReducer/mutationReducer
    isFetching: false,
    isProcessing: false,
    errors: new Map(),

    comments: new List(),
    metaTags: new Map(),
    pagination: new Map()
});

export default function commentReducer(state = new initState(), action) {
    switch (action.type) {
        case ActionTypes.COMMENT_FETCH_INIT:
        case ActionTypes.COMMENT_FETCH_SUCCESS:
        case ActionTypes.COMMENT_FETCH_ERROR:
            return fetchReducer(state, action, (payload) => ({
                comments: toList(payload.comments, Records.CommentRecord)
            }));

        case ActionTypes.COMMENT_CHANGE_INIT:
        case ActionTypes.COMMENT_CHANGE_SUCCESS:
        case ActionTypes.COMMENT_CHANGE_ERROR:
            return mutationReducer(state, action, (payload) => {
                let comments = state.comments;
                if (payload.deletedCommentIds) {
                    comments = state.comments.update(items => items.filter((item) => !payload.deletedCommentIds.includes(item.id)));
                } else {
                    const newItem = new Records.CommentRecord(payload.comment);
                    const index = findItemIndex(state.comments, newItem.id);
                    if (index === -1) {
                        if (newItem.parentId) {
                            const parentIndex = findItemIndex(state.comments, newItem.id, 'parentId');
                            comments = state.comments.insert(parentIndex, newItem);
                        } else {
                            comments = state.comments.update(items => items.concat([newItem]));
                        }
                    } else {
                        comments = state.comments.update(index, () => newItem);
                    }
                }

                return {
                    comments
                };
            }, ['comment' , 'deletedCommentIds']);

        default:
            return state;
    }
};
