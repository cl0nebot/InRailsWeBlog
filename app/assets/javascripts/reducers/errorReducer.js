'use strict';

import {
    Record,
    List,
    fromJS
} from 'immutable';

import * as ActionTypes from '../constants/actionTypes';

import {
    fetchReducer
} from './mutators';

const initState =  new Record({
    // Required for fetchReducer/mutationReducer
    isFetching: false,
    errors: new Map(),

    failures: new List(),
    pagination: new Map()
});

export default function conversationReducer(state = new initState(), action) {
    switch (action.type) {
        case ActionTypes.ERROR_FETCH_INIT:
        case ActionTypes.ERROR_FETCH_SUCCESS:
        case ActionTypes.ERROR_FETCH_ERROR:
            return fetchReducer(state, action, (payload) => ({
                failures: fromJS(payload.failures)
            }));

        case ActionTypes.ERROR_DELETE_SUCCESS:
            return state.merge({
                failures: state.failures.filter((failure) => action.removedId !== failure.get('id'))
            });
        case ActionTypes.ERROR_DELETE_ALL_SUCCESS:
            return state.merge({
                failures: new List()
            });

        default:
            return state;
    }
};
