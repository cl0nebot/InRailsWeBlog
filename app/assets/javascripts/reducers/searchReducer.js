'use strict';

import {
    Record,
    Map,
    List,
    fromJS
} from 'immutable';

import * as ActionTypes from '../constants/actionTypes';

import * as Records from '../constants/records';

import {
    toList,
    fetchReducer,
    addOrRemoveArray,
    mutateArray
} from './mutators';

// Autocomplete
const AutocompleteRecord = new Record({
    // Required for fetchReducer/mutationReducer
    isFetching: false,
    pagination: new Map(),
    errors: new Map(),

    query: undefined,

    actionKey: undefined,
    highlightedTagIndex: undefined,

    topics: new List(),
    tags: new List(),
    articles: new List()
});

export function autocompleteReducer(state = new AutocompleteRecord(), action) {
    switch (action.type) {
        case ActionTypes.SEARCH_AUTOCOMPLETE_FETCH_INIT:
        case ActionTypes.SEARCH_AUTOCOMPLETE_FETCH_SUCCESS:
        case ActionTypes.SEARCH_AUTOCOMPLETE_FETCH_ERROR:
            return fetchReducer(state, action, (payload) => ({
                query: payload.query,
                topics: toList(payload.topics, Records.TopicRecord),
                tags: toList(payload.tags, Records.TagRecord),
                articles: toList(payload.articles, Records.ArticleRecord)
            }));

        case ActionTypes.SEARCH_AUTOCOMPLETE_ACTION:
            return state.merge({
                actionKey: action.keyCode
            });

        case ActionTypes.SEARCH_TAG_SELECTED:
            return state.merge({
                query: '',
                actionKey: undefined
            });

        case ActionTypes.SEARCH_FETCH_SUCCESS:
            return state.merge({
                actionKey: undefined
            });

        default:
            return state;
    }
}

// Search
const SearchRecord = new Record({
    isSearching: false,
    hasResults: false,
    hasSearched: false,

    searchParams: new Map(),
    selectedTypes: new List(['article']), // initial query
    isShowingFilters: false,
    topicFilters: new Map(),
    topicActiveFilters: new Map(),
    tagFilters: new Map(),
    tagActiveFilters: new Map(),
    articleFilters: new Map(),
    articleActiveFilters: new Map(),

    query: undefined,

    selectedTags: new List(),

    topics: new List(),
    tags: new List(),
    articles: new List(),

    topicSuggestions: new List(),
    tagSuggestions: new List(),
    articleSuggestions: new List(),

    selectedItemType: undefined,
    selectedItemId: undefined,
    cancelSelection: false,
    selectedFromMap: false,

    totalArticlePages: undefined,
    totalArticles: undefined,
    totalTagPages: undefined,
    totalTags: undefined,
    totalTopicPages: undefined,
    totalTopics: undefined,
    currentPage: 1,

    relatedTopics: new List(),
    relatedTags: new List(),
    relatedArticles: new List(),

    metaResults: new Map()
});

const _parseSearchResults = (searchState, action) => {
    let newState = {};

    newState.isSearching = action.isSearching;
    newState.hasSearched = true;

    newState.query = action.query;

    newState.selectedTypes = new List(action.selectedTypes);
    newState.searchParams = fromJS(action.searchParams);

    newState.topics = toList(action.topics, Records.TopicRecord);
    newState.tags = toList(action.tags, Records.TagRecord);
    newState.articles = toList(action.articles, Records.ArticleRecord);

    newState.hasResults = !(newState.topics.isEmpty() && newState.tags.isEmpty() && newState.articles.isEmpty());

    newState.totalArticlePages = action.totalPages.articles;
    newState.totalArticles = action.totalCount.articles;
    newState.totalTagPages = action.totalPages.tags;
    newState.totalTags = action.totalCount.tags;
    newState.totalTopicPages = action.totalPages.topics;
    newState.totalTopics = action.totalCount.topics;

    newState.currentPage = parseInt(action.page || action.searchParams.page || 1, 10);

    if (action.suggestions) {
        newState.topicSuggestions = new List(action.suggestions.topics);
        newState.tagSuggestions = new List(action.suggestions.tags);
        newState.articleSuggestions = new List(action.suggestions.articles);
    }

    if (action.aggregations) {
        newState.topicFilters = new Map(action.aggregations.topics);
        newState.tagFilters = new Map(action.aggregations.tags);
        newState.articleFilters = new Map(action.aggregations.articles);
    }

    if (action.topicFilters) {
        newState.topicActiveFilters = searchState.topicActiveFilters.concat(action.topicFilters);
    }
    if (action.tagFilters) {
        newState.tagActiveFilters = searchState.tagActiveFilters.concat(action.tagFilters);
    }
    if (action.articleFilters) {
        newState.articleActiveFilters = searchState.articleActiveFilters.concat(action.articleFilters);
    }

    return searchState.merge(newState);
};

export function searchReducer(state = new SearchRecord(), action) {
    switch (action.type) {
        case ActionTypes.SEARCH_FETCH_INIT:
            return state.set('isSearching', action.isSearching);
        case ActionTypes.SEARCH_FETCH_ERROR:
            return state.merge({
                isSearching: false,
                hasResults: false
            });
        case ActionTypes.SEARCH_FETCH_SUCCESS:
            return _parseSearchResults(state, action);

        case ActionTypes.SEARCH_TAG_SELECTED:
            return state.merge({
                selectedTags: addOrRemoveArray(state.selectedTags, action.tag)
            });

        case ActionTypes.SEARCH_META_SUCCESS:
            return state.merge({
                isSearching: action.isSearching,
                query: action.query,
                metaResults: fromJS(action.metaResults)
            });

        // TOPIC
        case ActionTypes.TOPIC_CHANGE_SUCCESS:
            return state.merge({
                topics: mutateArray(state.topics, new Records.TopicRecord(action.topic), action.removedId)
            });

        // TAG
        case ActionTypes.TAG_CHANGE_SUCCESS:
            return state.merge({
                tags: mutateArray(state.tags, new Records.TagRecord(action.tag), action.removedId)
            });

        // ARTICLE
        case ActionTypes.ARTICLE_CHANGE_SUCCESS:
            return state.merge({
                articles: mutateArray(state.articles, new Records.ArticleRecord(action.article), action.removedId)
            });

        default:
            return state;
    }
}
