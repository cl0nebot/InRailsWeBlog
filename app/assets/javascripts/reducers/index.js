'use strict';

import {
    combineReducers
} from 'redux';

import {
    routerReducer
} from 'react-router-redux';

import uiReducer from './uiReducer';
import userReducer from './userReducer';
import articleReducer from './articleReducer';
import topicReducer from './topicReducer';
import tagReducer from './tagReducer';
import adminReducer from './adminReducer';

const rootReducer = combineReducers({
    router: routerReducer,
    uiState: uiReducer,
    userState: userReducer,
    articleState: articleReducer,
    topicState: topicReducer,
    tagState: tagReducer,
    adminReducer: adminReducer
});

export default rootReducer;
