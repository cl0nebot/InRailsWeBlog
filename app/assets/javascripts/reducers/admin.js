'use strict';

import {
    Record
} from 'immutable';

import {
    combineReducers
} from 'redux-immutable';

import uiReducer from './uiReducer';
import userReducer from './userReducer';
import adminReducer from './adminReducer';
import {
    autocompleteReducer,
    searchReducer
} from './searchReducer';
import topicReducer from './topicReducer';
import tagReducer from './tagReducer';
import articleReducer from './articleReducer';
import commentReducer from './commentReducer';
import errorReducer from './errorReducer';

const ReducerRecord = Record({
    uiState: undefined,
    userState: undefined,
    adminState: undefined,
    autocompleteState: undefined,
    searchState: undefined,
    topicState: undefined,
    tagState: undefined,
    articleState: undefined,
    commentState: undefined,
    errorState: undefined
});

const rootAdminReducer = combineReducers({
    uiState: uiReducer,
    userState: userReducer,
    adminState: adminReducer,
    autocompleteState: autocompleteReducer,
    searchState: searchReducer,
    topicState: topicReducer,
    tagState: tagReducer,
    articleState: articleReducer,
    commentState: commentReducer,
    errorState: errorReducer
}, ReducerRecord);

export default rootAdminReducer;
