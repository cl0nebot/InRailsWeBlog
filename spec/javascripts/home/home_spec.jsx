'use strict';

import '../test_helper';

import reducers from '../../../app/assets/javascripts/reducers';

import HomePage from '../../../app/assets/javascripts/components/home/home';

describe('<HomePage />', () => {
    let store;
    let shallowWrapper, mountWrapper;

    beforeEach(() => {
        // Mock errors reporting
        mock('/errors', 200, () => ({}));

        store = buildStore(reducers);

        jest.useFakeTimers();
    });

    it('renders home page', () => {
        shallowWrapper = shallow(<HomePage store={store}/>);

        expect(shallowWrapper.find('BrowserRouter')).toHaveLength(1);
        expect(shallowWrapper.find('Switch')).toHaveLength(1);
        expect(shallowWrapper.find('PasteManagerComponent(MatchMediaComponent(MainLayout))')).toBeTruthy();

        expect(shallowWrapper.dive()).toMatchSnapshot();
    });

    afterEach(() => {
        if (mountWrapper) {
            mountWrapper.unmount();
        }
    });
});
