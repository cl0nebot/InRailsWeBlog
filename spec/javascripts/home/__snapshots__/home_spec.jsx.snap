// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> renders home page 1`] = `
Array [
  <WithStyles(CssBaseline) />,
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <BrowserRouter>
      <withRouter(PasteManager)>
        <withRouter(ScrollBackManager)>
          <ErrorBoundary
            errorMessage="Une erreur est survenue lors de l'affichage du composant. Merci de recharger la page."
            errorTitle="Impossible d'afficher le menu"
            errorType="text"
          >
            <withRouter(Connect(WithStyles(HeaderLayoutHome)))
              permanentRoutes={
                Array [
                  Object {
                    "component": [Function],
                    "path": "search",
                  },
                ]
              }
            />
          </ErrorBoundary>
          <ErrorBoundary
            errorMessage="Une erreur est survenue lors de l'affichage du composant. Merci de recharger la page."
            errorTitle="Impossible d'afficher le composant"
            errorType="card"
          >
            <WithStyles(MainLayoutHome)
              permanentRoutes={
                Array [
                  Object {
                    "component": [Function],
                    "path": "new-article",
                  },
                  Object {
                    "component": [Function],
                    "path": "new-topic",
                  },
                ]
              }
              routes={
                Array [
                  Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/",
                  },
                  Object {
                    "component": [Function],
                    "hasBreadcrumb": false,
                    "path": "/search",
                  },
                  Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/tags/:tagSlug",
                  },
                  Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/tags",
                  },
                  Object {
                    "component": [Function],
                    "exact": false,
                    "path": "/tagged/:tagSlug/:childTagSlug?",
                  },
                  Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/users/:userSlug/topics/:topicSlug",
                  },
                  Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/users/:userSlug/topics/:topicSlug/tags",
                  },
                  Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/users/:userSlug",
                  },
                  Object {
                    "component": [Function],
                    "exact": true,
                    "path": "/users/:userSlug/articles/:articleSlug",
                  },
                  Object {
                    "component": [Function],
                    "path": "/404",
                  },
                ]
              }
            />
          </ErrorBoundary>
          <ErrorBoundary
            errorMessage="Une erreur est survenue lors de l'affichage du composant. Merci de recharger la page."
            errorTitle="Impossible d'afficher le bas de page"
            errorType="text"
          >
            <withRouter(WithStyles(FooterLayoutHome)) />
          </ErrorBoundary>
        </withRouter(ScrollBackManager)>
      </withRouter(PasteManager)>
    </BrowserRouter>
  </Provider>,
]
`;
